adapter/statsd2/statsd.go

```diff
@@ -0,0 +1,195 @@
+	"istio.io/mixer/template/metric"
+)
+
+const (
+	defaultFlushBytes = 512
+)
+
+type (
+	builder struct {
+		metrics map[string]*metric.Type
+	}
+
+	info struct {
+		mtype config.Params_MetricInfo_Type
+		tmpl  *template.Template
+	}
+
+	handler struct {
+		rate      float32
+		client    statsd.Statter
+		templates map[string]info // metric name -> template
+	}
+)
+
+var (
+	_ metric.HandlerBuilder = &builder{}
+	_ metric.Handler        = &handler{}
+)
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
+	return adapter.BuilderInfo{
+		Name:        "istio.io/mixer/adapter/statsd2",
+		Description: "Produces statsd metrics",
+		SupportedTemplates: []string{
+			metric.TemplateName,
+		},
+		DefaultConfig: &config.Params{
+			Address:       "localhost:8125",
+			Prefix:        "",
+			FlushDuration: 300 * time.Millisecond,
+			FlushBytes:    512,
+			SamplingRate:  1.0,
+		},
+		CreateHandlerBuilder: func() adapter.HandlerBuilder { return &builder{} },
+		ValidateConfig:       validateConfig,
+	}
+}
+
+func validateConfig(c adapter.Config) (ce *adapter.ConfigErrors) {
+	params := c.(*config.Params)
+	if params.FlushDuration < 0 {
+		ce = ce.Appendf("flushDuration", "flush duration must be >= 0")
+	}
+	if params.FlushBytes < 0 {
+		ce = ce.Appendf("flushBytes", "flush bytes must be >= 0")
+	}
+	if params.SamplingRate < 0 {
+		ce = ce.Appendf("samplingRate", "sampling rate must be >= 0")
+	}
+	for metricName, s := range params.Metrics {
+		if _, err := template.New(metricName).Parse(s.NameTemplate); err != nil {
+			ce = ce.Appendf("metricNameTemplateStrings", "failed to parse template '%s' for metric '%s': %v", s, metricName, err)
+		}
+	}
+	return
+}
+
+func (b *builder) ConfigureMetricHandler(metrics map[string]*metric.Type) error {
+	b.metrics = metrics
+	return nil
+}
+
+func (b *builder) Build(cfg adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	params := cfg.(*config.Params)
+
+	flushBytes := int(params.FlushBytes)
+	if flushBytes <= 0 {
+		env.Logger().Infof("Got FlushBytes of '%d', defaulting to '%d'", flushBytes, defaultFlushBytes)
+		// the statsd impl we use defaults to 1432 byte UDP packets when flushBytes <= 0; we want to default to 512 so we check ourselves.
+		flushBytes = defaultFlushBytes
+	}
+
+	client, _ := statsd.NewBufferedClient(params.Address, params.Prefix, params.FlushDuration, flushBytes)
+
+	templates := make(map[string]info)
+	for metricName, s := range params.Metrics {
+		def, found := b.metrics[metricName]
+		if !found {
+			env.Logger().Infof("template registered for nonexistent metric '%s'", metricName)
+			continue // we don't have a metric that corresponds to this template, skip processing it
+		}
+
+		var t *template.Template
+		if s.NameTemplate != "" {
+			t, _ = template.New(metricName).Parse(s.NameTemplate)
+			if err := t.Execute(ioutil.Discard, def.Dimensions); err != nil {
+				env.Logger().Warningf(
```

Comment:
Having mistyped stuff in the template seems like a very common/likely problem. Can we catch this in validateConfig instead so the user gets the error at the right place?

Reply:
We do already - lines 90-94. We can remove the check here entirely, but that felt a little off.